apps/studio/src/features/editing-experience/components/form-builder/renderers/controls/JsonFormsImageControl.tsx

```diff
@@ -0,0 +1,128 @@
+import type { ControlProps, RankedTester } from "@jsonforms/core"
+import { useEffect, useState } from "react"
+import { useParams } from "next/navigation"
+import { Box, FormControl, Text } from "@chakra-ui/react"
+import { and, isStringControl, rankWith, schemaMatches } from "@jsonforms/core"
+import { withJsonFormsControlProps } from "@jsonforms/react"
+import {
+  Attachment,
+  FormErrorMessage,
+  FormLabel,
+} from "@opengovsg/design-system-react"
+import wretch from "wretch"
+
+import { JSON_FORMS_RANKING } from "~/constants/formBuilder"
+import { trpc } from "~/utils/trpc"
+import {
+  IMAGE_UPLOAD_ACCEPTED_MIME_TYPES,
+  MAX_IMG_FILE_SIZE_BYTES,
+} from "./constants"
+
+export const jsonFormsImageControlTester: RankedTester = rankWith(
+  JSON_FORMS_RANKING.ImageControl,
+  and(
+    isStringControl,
+    schemaMatches((schema) => schema.format === "image"),
+  ),
+)
+export function JsonFormsImageControl({
+  data,
+  label,
+  handleChange,
+  path,
+  description,
+  required,
+}: ControlProps) {
+  const { pageId, siteId } = useParams()
+
+  const [selectedFile, setSelectedFile] = useState<File | undefined>()
+  const [pendingFile, setPendingFile] = useState<File | undefined>()
+  const [errorMessage, setErrorMessage] = useState("")
+
+  useEffect(() => {
+    if (!!data) {
+      wretch(data as string)
+        .get()
+        .blob()
+        .then((blob) => {
+          const splitData = (data as string).split("/")
+          const fileName = splitData[-1] || "Current Image"
+          setSelectedFile(new File([blob], fileName))
+        })
+        .catch((error) => {
+          console.log("error fetching initial image", error)
+        })
+    }
+    // NOTE: Using empty dependency array because we are checking if fetch is needed only upon initial load.
+    // After this load, we are the only editor of this page, and any image url changes are caused by us and we will be caching the file locally.
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [])
+  const getPresignedMutation =
+    trpc.page.getPresignUrlForImageUpload.useMutation()
+  const uploadImage = async (image: File) => {
+    const { presignedUploadURL, fileURL } =
+      await getPresignedMutation.mutateAsync({
+        pageId: Number(pageId),
+        siteId: Number(siteId),
+      })
+    const response = await wretch(presignedUploadURL)
+      .content(image.type)
+      .put(image)
+      .res()
+    if (response.ok) {
+      setSelectedFile(pendingFile)
+      setErrorMessage("")
+      handleChange(path, fileURL)
+      console.log("new file url", fileURL)
+    } else {
+      setPendingFile(undefined)
+      setErrorMessage(
```

Comment:
is this a toast?

Reply:
This isn't a toast, but it's a FormErrorMessage. Not too sure about going with a toast, since non of the other renderers are using it at the moment for error flows, so I'm wary that it might be out of place.